posts/2021/03/array_interchange_protocols.md

```diff
@@ -0,0 +1,370 @@
+In Python, the data buffers of extension types can be accessed using `memoryview` object.
+
+
+### NumPy's `__array__` protocol
+
+NumPy ndarrays can be constructed from objects that implement
+[`__array__([dtype])`
+method](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.__array__.html#numpy.ndarray.__array__). The
+`__array__` method returns a NumPy ndarray that may or [may
+not](https://numpy.org/doc/stable/user/basics.dispatch.html) be a view
+of the object's internal data buffer structure.
+
+
+### DLPack
+
+All the above mentioned protocols define sharing the data buffers of
+array-like objects without specifying a device where the raw-pointer
+values are valid addresses of device memory. The device is assumed to
+be fixed: CPU for the content of `__array_interface__`,
+`__array_struct__`, `__array__()`, and `memoryview` structures; CUDA
+for `__cuda_array_interface__`. With the emergence of various
+processor devices for solving HPC/ML problems, these protocols are
+essentially flawed in the systems with heterogeneous processor units.
+
+[DLPack](https://github.com/dmlc/dlpack/) introduces a in-memory
+structure for sharing the data buffers in a device aware way. For
+instance, the structure can currently hold the data buffers of the
+following device types: CPU, CUDA GPU, OpenCL, Vulkan GPU, Verilog,
+and ROCm GPU.
+
+## Supported features
+
+TODO: this may be better as separate tables, or with more/less features listed:
+
+| *supports*                 | Python API | C API | CPU | CUDA | implementation independent | uint, int, float, complex | string | datetime | PyObject | bfloat16 |
+|----------------------------|:----------:|:-----:|:---:|:----:|----------------------------|---------------------------|--------|----------|----------|----------|
+| buffer protocol            |            |   Y   |  Y  |      | Y                          | Y                         | Y      | Y        | Y        |          |
+| `__array_interface__`      |      Y     |       |  Y  |      | Y                          | Y                         | Y      | Y        | Y        |          |
+| `__array_struct__`         |      Y     |   Y   |  Y  |      | Y                          | Y                         | Y      | Y        | Y        |          |
+| `__cuda_array_interface__` |      Y     |       |     |   Y  | Y                          | Y                         | Y      | Y        | Y        |          |
+| `__array__`                |            |       |  Y  |      |                            | Y                         |        | Y        | Y        |          |
+| DLPack                     |      Y     |   Y   |  Y  |   Y  | Y                          | Y                         |        |          |          | Y        |
+
+TODO: an alternative view of the above table:
+
+| *Protocols:*              | buffer protocol  | `__array_interface__` | `__array_struct__` | `__cuda_array_interface__` | NumPy `__array__()` | DLPack |
```

Comment:
This is probably a better view indeed

Reply:
In github view of a Markdown document this looks better, but in deploy-preview both tables look ugly to me.
The third option would be to provide the table as an (svg) image.

Reply:
Agreed, pretty ugly - an SVG image sounds like a great idea.